### PR TITLE
Print link to project reports on CLI

### PR DESCRIPTION
**Pull request by johnmartel**
Any of the checked boxes below indicate that I took action:
- [x] Reviewed the [Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md#contribution-workflow).
- [x] Signed the [Gradle CLA](http://gradle.org/contributor-license-agreement/).
- [x] Ensured that basic checks pass: `./gradlew quickCheck`

For all non-trivial changes that modify the behavior or public API:
- [ ] Before submitting a pull request, I started a discussion on the [Gradle developer list](https://groups.google.com/forum/#!forum/gradle-dev),
    the [forum](https://discuss.gradle.org/) or can reference a [JIRA issue](https://issues.gradle.org/secure/Dashboard.jspa).
- [ ] I considered writing a design document. A design document can be
  brief but explains the use case or problem you are trying to solve,
  touches on the planned implementation approach as well as the test cases
  that verify the behavior. Optimally, design documents should be submitted
  as a separate pull request. [Samples](https://github.com/gradle/gradle/tree/master/design-docs)
  can be found in the Gradle GitHub repository. Please let us know if you need help with
  creating the design document. We are happy to help!
- [x] The pull request contains an appropriate level of unit and integration
  test coverage to verify the behavior. Before submitting the pull request
  I ran a build on your local machine via the command
  `./gradlew quickCheck <impacted-subproject>:check`.
- [ ] The pull request updates the Gradle documentation like user guide,
  DSL reference and Javadocs where applicable.

Fixes gradle/gradle#712

### CODE REVIEW

File: subprojects/diagnostics/src/integTest/groovy/org/gradle/api/plugins/ProjectReportsPluginIntegrationTest.java [hunk 1]
```diff
@@ -16,18 +16,85 @@
...
         testFile("build/reports/project/tasks.txt").assertExists();
     }
+
+    private void applyProjectReportPlugin() {
+        testFile("build.gradle").writelns(
+            "apply plugin: 'project-report'"
+        );
+    }
+
+    @Test
```
Comment (johnmartel) [inline subprojects/diagnostics/src/integTest/groovy/org/gradle/api/plugins/ProjectReportsPluginIntegrationTest.java]: There is some test duplication starting here, I don't know how you guys prefer it, that could easily be extracted to a method that receives the task name and the expected path as parameters.

File: subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/AbstractReportTask.java [hunk 2]
```diff
@@ -71,6 +72,10 @@ public void generate() {
             @Override
             public void generateReport(Project project) throws IOException {
                 generate(project);
+
+                if (!outputsToConsole()) {
```
Comment (johnmartel) [inline subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/AbstractReportTask.java]: Could also  reverse the condition and name the private method to outputsToFile() instead.

**Comment by johnmartel** [comment]
I am missing unit tests for `AbstractReportTask`. I am wondering how to best test the interaction with the `project.logger` in `AbstractReportTaskTest` since I can't assign the logger and it is not a mock. Any hints would be appreciated, as I normally would have written a unit test even before making that change.

**Comment by johnmartel** [comment]
Tried to use `project.getLogging().addStandardOutputListener()` to no avail:

```java
    @Test
    public void logsLinkToReportFile() throws IOException {
        final File outputFile = tmpDir.getTestDirectory().file("build/reports/project/dependencies.txt");
        MatchingStandardOutputListener logListener = new MatchingStandardOutputListener(TextUtil.normaliseFileSeparators(outputFile.getAbsolutePath()));

        context.checking(new Expectations() {{
            Sequence sequence = context.sequence("sequence");
            one(renderer).setClientMetaData((BuildClientMetaData) with(notNullValue()));
            inSequence(sequence);
            one(renderer).setOutputFile(outputFile);
            inSequence(sequence);
            one(renderer).startProject(project);
            inSequence(sequence);
            one(generator).run();
            inSequence(sequence);
            one(renderer).completeProject(project);
            inSequence(sequence);
            one(renderer).complete();
            inSequence(sequence);
        }});

        project.getLogging().addStandardOutputListener(logListener);
        task.setOutputFile(outputFile);
        task.generate();

        assertThat(logListener.matchedExpectedOutput, is(true));
    }

    private static class MatchingStandardOutputListener implements StandardOutputListener {
        private final String expectedOutput;
        private boolean matchedExpectedOutput;

        MatchingStandardOutputListener(String expectedOutput) {
            this.expectedOutput = expectedOutput;
        }

        @Override
        public void onOutput(CharSequence output) {
            matchedExpectedOutput = matchedExpectedOutput || StringUtils.contains(output, expectedOutput);
        }
    }
```

When invoking `project.getLogger().lifecycle()`, my StandardOutputListener is not called.

**Comment by big-guy** [comment]
Don't kill yourself over trying to make a test in `AbstractReportTaskTest`.  It's an ancient JMock test that we've been slowly getting rid of/converting.

I think an integration test is fine for this kind of functionality.  The important part is what the user sees.

File: subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/AbstractReportTask.java [hunk 3]
```diff
@@ -71,6 +72,10 @@ public void generate() {
...
+                    project.getLogger().lifecycle("See the report at: file://{}", TextUtil.normaliseFileSeparators(getOutputFile().getAbsolutePath()));
```
Comment (big-guy) [inline subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/AbstractReportTask.java]: This should use `org.gradle.internal.logging.ConsoleRenderer#asClickableFileUrl`

File: subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/AbstractReportTask.java [hunk 4]
```diff
@@ -79,6 +84,10 @@ public void generateReport(Project project) throws IOException {
         reportGenerator.generateReport(new TreeSet<Project>(getProjects()));
     }
 
+    private boolean outputsToConsole() {
```
Comment (big-guy) [inline subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/AbstractReportTask.java]: I'd probably use a name like `shouldCreateReportFile`

File: subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/AbstractReportTask.java (hunk 2 above)
Comment (big-guy) [inline subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/AbstractReportTask.java]: Yes, I'd probably reverse it and make this check more like "are we going to have a report file?" (I suggested a different name below).

File: subprojects/diagnostics/src/integTest/groovy/org/gradle/api/plugins/ProjectReportsPluginIntegrationTest.java (hunk 1 above)
Comment (big-guy) [inline subprojects/diagnostics/src/integTest/groovy/org/gradle/api/plugins/ProjectReportsPluginIntegrationTest.java]: We prefer Spock tests, but we have a mix of older test classes (like this one) that don't use it.  We usually convert tests as we find them, but I don't want to ask you to do that. I went ahead and converted it: https://github.com/gradle/gradle/blob/master/subprojects/diagnostics/src/integTest/groovy/org/gradle/api/plugins/ProjectReportsPluginIntegrationSpec.groovy

If you wouldn't mind deleting this class and putting your tests into `ProjectReportsPluginIntegrationSpec` that would be appreciated.  Having a parameterized test sounds fine.  I don't think we have to assert on the exact message (with the full path).  Just asserting that the output contains "See the report at:" is probably enough (that's what we do elsewhere).

Comment (johnmartel) [inline subprojects/diagnostics/src/integTest/groovy/org/gradle/api/plugins/ProjectReportsPluginIntegrationTest.java]: Ok, I did not know about the refactoring to Spock tests, although I noted the mix of JUnit and Spock. I'll move the tests to the specification class and will do the same in the future.

Thanks for clarifying that.

**Comment by eriwen** [comment]
@johnmartel For future contributions, [allowing maintainers to make changes to the submitted branch](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) would streamline things like the Spock conversion. 

Not required, just a suggestion :) /cc @big-guy

**Comment by johnmartel** [comment]
These last commits complete this PR and fix gradle/gradle#712.

File: subprojects/diagnostics/src/integTest/groovy/org/gradle/api/plugins/ProjectReportsPluginIntegrationSpec.groovy [hunk 5]
```diff
@@ -28,22 +29,39 @@ class ProjectReportsPluginIntegrationSpec extends AbstractIntegrationSpec {
...
+    def "prints link to default #task"(String task) {
+        when:
+        succeeds(task)
+
+        then:
+        outputContains("See the report at:")
+
+        where:
+        task << ["taskReport", "propertyReport", "dependencyReport", "htmlDependencyReport"]
     }
```
Comment (big-guy) [inline subprojects/diagnostics/src/integTest/groovy/org/gradle/api/plugins/ProjectReportsPluginIntegrationSpec.groovy]: Could you also provide a test that explicitly sets outputFile = null for these tasks and asserts that the output does not contain the "See the report at:" message?

**Comment by johnmartel** [comment]
Good point, I added the missing tests.

**Comment by big-guy** [comment]
This'll be in 3.3.  Thanks!

**Pull request merged**